agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -1268,49 +1261,83 @@ bool backhaul_manager::send_autoconfig_search_message(std::shared_ptr<sRadioInfo
+            return false;
+        }
 
-                if (!query->alloc_bssid_list(1)) {
-                    LOG(ERROR) << "Failed allocate memory for bssid_list";
-                    return false;
-                }
+        // copy current vaps bssid
+        std::vector<sMacAddr> current_socket_bssids;
+        std::transform(socket->vaps_list.vaps, socket->vaps_list.vaps + beerocks::IFACE_TOTAL_VAPS,
+                       std::back_inserter(current_socket_bssids),
+                       [](const beerocks_message::sVapInfo &vap_info) { return vap_info.mac; });
 
-                auto list         = query->bssid_list(0);
-                std::get<0>(list) = true;
-                std::get<1>(list) = mac;
+        // sort current vaps bssid
+        std::sort(current_socket_bssids.begin(), current_socket_bssids.end(), mac_comp);
 
-                if (!message_com::send_cmdu(socket->slave, cmdu_tx)) {
-                    LOG(ERROR) << "Failed forwarding AP_METRICS_QUERY_MESSAGE message to son_slave";
-                    ret = false;
-                    continue;
-                } else {
-                    ret = true;
-                    // Fill a query vector
-                    m_ap_metric_query.push_back({socket->slave, mac});
-                }
-            }
+        // fill a list of the common bssids: given list with current vaps
+        std::vector<sMacAddr> required_bssids;
+        std::set_intersection(given_bssid_list.begin(), given_bssid_list.end(),
+                              current_socket_bssids.begin(), current_socket_bssids.end(),
+                              std::back_inserter(required_bssids), mac_comp);
+
+        if (required_bssids.empty()) {
+            // skip if there is nothing to do
+            LOG(DEBUG) << "empty required bssid list, skipping";
+            continue;
+        }
+
+        // add metric query tlv
+        auto query = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
+        if (!query) {
+            LOG(ERROR) << "Failed addClass<wfa_map::tlvApMetricQuery>";
+            return false;
+        }
+
+        // alocate memory
+        if (!query->alloc_bssid_list(required_bssids.size())) {
+            LOG(ERROR) << "Failed allocate memory for bssid_list, size: " << bssid_list.size();
+            return false;
+        }
+
+        int i = 0;
+        for (const auto &mac : required_bssids) {
+            auto list         = query->bssid_list(i);
+            std::get<0>(list) = true;
+            std::get<1>(list) = mac;
             i++;
         }
+
+        // save the required list
+        m_expected_ap_metrics_response.add_expected_bssid(required_bssids.begin(),
+                                                          required_bssids.end());
+
+        if (!message_com::send_cmdu(socket->slave, cmdu_tx)) {
+            LOG(ERROR) << "Failed forwarding AP_METRICS_QUERY_MESSAGE message to son_slave";
+            ret &= false;
```

Comment:
I don't know. Technically you are right. However, we might change the response implementation in the future, to send back "whatever was returned". In this case stopping on first fail would be wrong.
Moreover, this loop creates an asynchronous process of requests. Stopping on failure may result in different runs just because the order of the the sockets is different.
All in all, I am not sure that stopping on error is correct here.